### PR TITLE
CDAP-3625: Improved table naming for stream views

### PR DESCRIPTION
**Pull request by jwang47**
- "tableName" can now be passed into the HTTP endpoint for
  creating/updating a stream view so the user can specify what
  the name of the Hive table will be
- default table name for stream views changed from
  `streams_<stream-id>_views_<view-id>` to
  `stream_<stream-id>_<view_id>`

https://issues.cask.co/browse/CDAP-3625
http://builds.cask.co/browse/CDAP-DUT2775

### CODE REVIEW

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/CreateOrUpdateStreamViewCommand.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
       settings = ArgumentParser.parseMap(arguments.get(ArgumentName.SETTINGS.toString()));
     }
     FormatSpecification formatSpecification = new FormatSpecification(formatName, schema, settings);
-    ViewSpecification viewSpecification = new ViewSpecification(formatSpecification);
+    ViewSpecification viewSpecification = new ViewSpecification(formatSpecification, null);
```
Comment (chtyim) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/CreateOrUpdateStreamViewCommand.java]: Can you have an overloaded constructor so that passing in `null` in not required?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/DescribeStreamViewCommand.java [hunk 2]
```diff
@@ -59,11 +59,12 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
     ViewDetail detail = client.get(Id.Stream.View.from(streamId, arguments.get(ArgumentName.VIEW.toString())));
 
     Table table = Table.builder()
-      .setHeader("id", "format", "schema", "settings")
+      .setHeader("id", "format", "table", "schema", "settings")
       .setRows(Collections.singletonList(detail), new RowMaker<ViewDetail>() {
         @Override
         public List<?> makeRow(ViewDetail object) {
           return Lists.newArrayList(object.getId(), object.getFormat().getName(),
+                                    object.getTableName(),
```
Comment (chtyim) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/DescribeStreamViewCommand.java]: Will it show `null` or always show the actual explore table name?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java [hunk 3]
```diff
@@ -43,9 +43,17 @@ public ViewAdmin(ViewStore store, ExploreFacade explore, ExploreTableNaming nami
 
   public boolean createOrUpdate(Id.Stream.View viewId, ViewSpecification spec) throws Exception {
     if (store.exists(viewId)) {
-      explore.disableExploreStream(viewId.getStream(), naming.getTableName(viewId));
+      ViewSpecification previousSpec = store.get(viewId);
+      if (previousSpec == null) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: Do you still need to check for `exists` then? Isn't `null` previousSpec means it does not exist?

File: cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/DescribeStreamViewCommand.java (hunk 2 above)
Comment (jwang47) [inline cdap-cli/src/main/java/co/cask/cdap/cli/command/stream/view/DescribeStreamViewCommand.java]: Always the actual table name. Table name is only nullable in client code or `ViewAdmin.createOrUpdate`. A persisted `ViewSpecification` always has an actual table name.

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java [hunk 4]
```diff
@@ -126,7 +126,8 @@ public QueryHandle enableStream(String tableName, Id.Stream streamID, FormatSpec
       .setTableComment("CDAP Stream")
       .buildWithStorageHandler(StreamStorageHandler.class.getName(), serdeProperties);
 
-    LOG.debug("Running create statement for stream {} with table {}: {}", streamName, tableName, createStatement);
+    // change back to debug
```
Comment (chtyim) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java]: So why the change?

Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java]: Oops, forgot to change back.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java [hunk 5]
```diff
@@ -42,19 +42,26 @@ public ViewAdmin(ViewStore store, ExploreFacade explore, ExploreTableNaming nami
   }
 
   public boolean createOrUpdate(Id.Stream.View viewId, ViewSpecification spec) throws Exception {
-    if (store.exists(viewId)) {
-      explore.disableExploreStream(viewId.getStream(), naming.getTableName(viewId));
+    try {
+      ViewSpecification previousSpec = store.get(viewId);
+      if (previousSpec == null) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: Is this possible? What's the contract of the `store.get` method?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: Asking because it seems like it will always throw `NotFoundException` and the `store.get` is not nullable.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java [hunk 6]
```diff
@@ -42,19 +42,26 @@ public ViewAdmin(ViewStore store, ExploreFacade explore, ExploreTableNaming nami
...
+        throw new IllegalStateException("View exists, but no specification found");
+      }
+      explore.disableExploreStream(viewId.getStream(), previousSpec.getTableName());
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: So for updating, we actually allow changing the explore table name as well? Is it a good?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java (hunk 5 above)
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: Oops. It's not possible.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java (hunk 6 above)
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: Could be a bit confusing.

File: cdap-proto/src/main/java/co/cask/cdap/proto/ViewSpecification.java [hunk 7]
```diff
@@ -26,18 +27,34 @@
...
+  public ViewSpecification(FormatSpecification format, String tableName) {
     this.format = format;
+    this.tableName = tableName;
+  }
+
+  public ViewSpecification(FormatSpecification format) {
+    this(format, null);
   }
 
   public ViewSpecification(ViewSpecification existing) {
```
Comment (chtyim) [inline cdap-proto/src/main/java/co/cask/cdap/proto/ViewSpecification.java]: nit: Better call it `other`.

**Comment by chtyim** [comment]
Just one minor comment. Rest LGTM.
